src/main/surfrdUtilsMod.F90

```diff
@@ -143,6 +145,140 @@ subroutine convert_cft_to_pft( begg, endg, cftsize, wt_cft )
 
   end subroutine convert_cft_to_pft
 
+  !-----------------------------------------------------------------------
+  subroutine collapse_nat_pfts(wt_nat_patch, natpft_size, begg, endg, n_dom_pfts)
+    !
+    ! DESCRIPTION
+    ! Collapse to the top N dominant pfts (n_dom_pfts)
+    ! - Bare ground could be up to 1 patch before collapsing.
+    ! - Pfts could be up to 14 before collapsing if create_crop_landunit = .T.
+    ! - Pfts could be up to 16 before collapsing if create_crop_landunit = .F.
+    !
+    ! !USES:
+    use clm_varpar, only: natpft_lb, natpft_ub, maxveg
+    use pftconMod, only: noveg, nc3crop, nc3irrig, pftcon
+    use array_utils, only: find_k_max_indices
+    !
+    ! !ARGUMENTS:
+    ! Use begg and endg rather than 'bounds', because bounds may not be
+    ! available yet when this is called
+    integer, intent(in) :: begg  ! Beginning grid cell index
+    integer, intent(in) :: endg  ! Ending grid cell index
+    integer, intent(in) :: natpft_size  ! how many pfts 
+    integer, intent(in) :: n_dom_pfts  ! # dominant soil pfts
+    ! These arrays modified in-place
+    ! Weights of pfts per grid cell
+    ! Dimensioned [g, natpft_lb:natpft_ub]
+    real(r8), intent(inout) :: wt_nat_patch(begg:, natpft_lb:)
+    !
+    ! !LOCAL VARIABLES:
+    integer :: g  ! gridcell index
+    integer :: m  ! pft index
+    integer :: n  ! index of the order of the dominant pfts
+    integer, allocatable :: max_indices(:)  ! array of dominant pft index values
+    real(r8) :: wt_dom_nat_sum
+
+    character(len=*), parameter :: subname = 'collapse_nat_pfts'
+
+    !-----------------------------------------------------------------------
+
+    SHR_ASSERT_ALL((ubound(wt_nat_patch) == (/endg, natpft_lb+natpft_size-1/)), errMsg(sourcefile, __LINE__))
+
+    ! Find the top N dominant pfts to collapse the data to
+    ! n_dom_pfts < 0 is not allowed (error check in controlMod.F90)
+    ! Default value n_dom_pfts = 0 or a user-selected n_dom_pfts = natpft_size
+    ! means "do not collapse pfts" and skip over this subroutine's work
+    if (n_dom_pfts > 0 .and. n_dom_pfts < natpft_size) then
+       allocate(max_indices(n_dom_pfts))
+       do g = begg, endg
+          max_indices = 0._r8  ! initialize
+          call find_k_max_indices(wt_nat_patch(g,:), natpft_lb, n_dom_pfts, &
+                                  max_indices)
+
+          ! Adjust wt_nat_patch by normalizing the dominant weights to 1
+          ! (currently they sum to <= 1) and setting the remaining weights to 0.
+          ! TODO Possible to use existing function renormalize?
+          wt_dom_nat_sum = 0._r8  ! initialize the dominant pft sum
+          do n = 1, n_dom_pfts
+             m = max_indices(n)
+             wt_dom_nat_sum = wt_nat_patch(g,m) + wt_dom_nat_sum
+          end do
+          ! Normalize dominant pft weights to 1; if non-existent,
+          ! set the pft weights and pft landunit weight to 0.
+          if (wt_dom_nat_sum <= 0._r8) then
```

Comment:
@slevisconsulting OK this case could only happen if the n_dom_pfts are all zeroed. Which basically means that all PFTS add up to zero. So that could only be if PCT_NATVEG=0 for that point. I suppose what you are doing here is harmless, just saying that everything is in bare-soil. But, it's handling a special case that maybe it shouldn't? I'm also wondering about the allowance for wt_dom_nat_sum to be less than zero, as that indicates you have weights that are negative, and I'm not sure we should allow that? But, I can't say offhand how the code handles negative weights elsewhere.

@billsacks do you have any thoughts on this? It's doing something that's harmless, and as is it might be needed for cases where PCT_NATVEG=0. But, you could also just not do the collapse for those cases.

Reply:
I agree with @ekluzek 's sense here. But I'll go further: If I remember correctly, `wt_nat_patch` will always add to 1, even if PCT_NATVEG = 0. (This is needed for dynamic landunits.) So I'd keep this `if (wt_dom_nat_sum <= 0._r8)` conditional, but delete the current body and put an endrun call in there instead: if the code is working right, I think we should _always_ have `wt_dom_nat_sum > 0`, and that would be a good thing to confirm.

Actually, that depends on assuming that `n_dom_pfts >= 1`; you could put an assertion of that at the top of the routine (but it's not critical).